crates/bevy_reflect/src/serde/de/deserializer.rs

```diff
@@ -14,6 +14,157 @@ use crate::{
+/// # struct LoadedUntypedAsset;
+/// # #[derive(Debug, Clone, Reflect)]
+/// # struct Handle<T: Reflect>(T);
+/// # #[derive(Debug, Clone, Reflect)]
+/// # struct Mesh;
+/// #
+/// # struct LoadContext;
+/// # impl LoadContext {
+/// #     fn load(&mut self) -> &mut Self { unimplemented!() }
+/// #     fn with_asset_type_id(&mut self, (): ()) -> &mut Self { unimplemented!() }
+/// #     fn untyped(&mut self) -> &mut Self { unimplemented!() }
+/// #     fn load_asset(&mut self, (): ()) -> Handle<LoadedUntypedAsset> { unimplemented!() }
+/// # }
+/// #
+/// # struct ReflectHandle;
+/// # impl TypeData for ReflectHandle {
+/// #     fn clone_type_data(&self) -> Box<dyn TypeData> { unimplemented!() }
+/// # }
+/// # impl ReflectHandle {
+/// #     fn asset_type_id(&self) { unimplemented!() }
+/// # }
+/// #
+/// #[derive(Debug, Clone, Reflect)]
+/// struct MyAsset {
+///     name: String,
+///     mesh: Handle<Mesh>,
+/// }
+///
+/// # type AssetError = Box<dyn core::error::Error>;
+/// # fn load<AssetPathVisitor: for<'de> Visitor<'de, Value = ()> + Copy>(
+/// #     asset_bytes: &[u8],
+/// #     type_registry: &TypeRegistry,
+/// #     load_context: &mut LoadContext,
+/// #     AssetPathVisitor: AssetPathVisitor,
+/// # ) -> Result<MyAsset, AssetError> {
+/// let mut ron_deserializer = ron::Deserializer::from_bytes(asset_bytes)?;
+/// let mut processor = ReflectDeserializerProcessor::new(
+///     |registration: &TypeRegistration| registration.data::<ReflectHandle>().is_some(),
+///     |registration, deserializer| {
+///         let reflect_handle = registration.data::<ReflectHandle>().unwrap();
+///         let asset_type_id = reflect_handle.asset_type_id();
+///
+///         let asset_path = deserializer.deserialize_str(AssetPathVisitor)?;
+///
+///         // takes in `load_context` from its scope
+///         let handle: Handle<LoadedUntypedAsset> = load_context
+///             .load()
+///             .with_asset_type_id(asset_type_id)
+///             .untyped()
+///             .load_asset(asset_path);
+/// #       let _: Result<_, ()> = {
+///         Ok(Box::new(handle))
+/// #       };
+/// #       unimplemented!()
+///     },
+/// );
+/// let reflect_deserializer =
+///     ReflectDeserializer::new_with_processor(type_registry, &mut processor);
+/// let asset = reflect_deserializer.deserialize(&mut ron_deserializer)?;
+/// # unimplemented!()
+/// # }
+/// ```
+///
+/// [`can_deserialize`]: Self::can_deserialize
+/// [`deserialize`]: Self::deserialize
+pub struct ReflectDeserializerProcessor<'p> {
+    /// When deserializing a value of which we know the type (via
+    /// [`TypeRegistration`]), do we want this processor to override the default
+    /// deserialization?
+    ///
+    /// If this returns [`true`], [`deserialize`] is used to create the
+    /// reflected value.
+    ///
+    /// [`deserialize`]: Self::deserialize
+    pub can_deserialize: Box<dyn FnMut(&TypeRegistration) -> bool + 'p>,
```

Comment:
This should probably return a `Result`. If we are given a type, `Foo`, that we know must be deserialized via the processor, but is missing certain type data or other criteria, then we may want to return an error right then and there.

Currently, you have to either return `true` and generate the error in `deserialize` or return `false` and have the deserialization fail with a less specific error message.

Reply:
I agree, this would be a nice API feature. Ideally, we just merge `should_deserialize` and `deserialize`, but this causes problems with the `erased_serde::Deserializer` (https://github.com/bevyengine/bevy/pull/15482#issuecomment-2380259729).
I'll change the sig to `-> Result`.